wrappers/services.go

```diff
@@ -471,116 +471,178 @@ type EnsureSnapServicesOptions struct {
 	// are any issues.
 	// note that the rollback is best effort, if we are rebooted in the middle,
 	// there is no guarantee about the state of files, some may have been
 	// updated and some may have been rolled back, higher level tasks/changes
 	// should have do/undo handlers to properly handle the case where this
 	// function is interrupted midway
-	modifiedUnitsPreviousState := make(map[string]*osutil.MemoryFileState)
-	var modifiedSystem, modifiedUser bool
+	modifiedFiles map[string]*osutil.MemoryFileState
+}
 
-	defer func() {
-		if err == nil {
-			return
+// restore is a helper function which should be called in case any errors happen
+// during the write/update of systemd files
+func (es ensureSnapServicesContext) restore() {
+	for file, state := range es.modifiedFiles {
+		if state == nil {
+			// we don't have anything to rollback to, so just remove the
+			// file
+			if err := os.Remove(file); err != nil {
+				es.inter.Notify(fmt.Sprintf("while trying to remove %s due to previous failure: %v", file, err))
+			}
+		} else {
+			// rollback the file to the previous state
+			if err := osutil.EnsureFileState(file, state); err != nil {
+				es.inter.Notify(fmt.Sprintf("while trying to rollback %s due to previous failure: %v", file, err))
+			}
 		}
-		for file, state := range modifiedUnitsPreviousState {
-			if state == nil {
-				// we don't have anything to rollback to, so just remove the
-				// file
-				if e := os.Remove(file); e != nil {
-					inter.Notify(fmt.Sprintf("while trying to remove %s due to previous failure: %v", file, e))
-				}
-			} else {
-				// rollback the file to the previous state
-				if e := osutil.EnsureFileState(file, state); e != nil {
-					inter.Notify(fmt.Sprintf("while trying to rollback %s due to previous failure: %v", file, e))
-				}
+	}
+
+	if !es.opts.Preseeding {
+		if es.modifiedSystem {
+			if err := es.sysd.DaemonReload(); err != nil {
+				es.inter.Notify(fmt.Sprintf("while trying to perform systemd daemon-reload due to previous failure: %v", err))
 			}
 		}
-		if modifiedSystem && !preseeding {
-			if e := sysd.DaemonReload(); e != nil {
-				inter.Notify(fmt.Sprintf("while trying to perform systemd daemon-reload due to previous failure: %v", e))
+		if es.modifiedUser {
+			if err := userDaemonReload(); err != nil {
+				es.inter.Notify(fmt.Sprintf("while trying to perform user systemd daemon-reload due to previous failure: %v", err))
 			}
 		}
-		if modifiedUser && !preseeding {
-			if e := userDaemonReload(); e != nil {
-				inter.Notify(fmt.Sprintf("while trying to perform user systemd daemon-reload due to previous failure: %v", e))
+	}
+}
+
+// reloadModified uses the modifiedSystem/modifiedUser to determine whether a reload
+// is required from systemd to take the new systemd files into effect. This is a NOP
+// if preseeding is set in opts.Preseeding
+func (es ensureSnapServicesContext) reloadModified() error {
+	if !es.opts.Preseeding {
```

Comment:
Nitpick, I think this condition could be reversed to have less indentation:
```
if es.opts.Preseeding {
    return
}
```

Reply:
Ah nice catch, fixed!